### DISCUSSION TITLE
Making notcurses happy!

### DISCUSSION THREAD

**Discussion by christianparpart** [contour-terminal/contour#444] [Development]
Hey @dankamongmen (why is autocompletion working for everyone but not for your name? so sad I am).

In my attempt to make notcurses as happy as it can get  - simply by looking at the output of `notcurses-info` for the time being I realized that:

* you don't seem to believe that Contour can do undercurl, but we support various styled underlines as well as colorizing them (de-facto std CSI sequences)
* `3x2-` is indicated as not available. hm, I agree, I just merged these to master.
* `defbg 0x101010 not considered transparent` How do you decide that something is or is not transparent? Am I maybe not passing all the information to you so that you can find out?
* smooth mosaics: I will be working on them next
* mathematical symbols (I so far only implemented square brackets I think, yesterday, rest will follow)
* you detect that sixels are supported but I do not see the tiny graphics in the top right anymore (was it ever graphics? I don't remember, but that fancy logo is missing)
* roundish box drawings and diagonal mosaics I will be adding soon.
* **emoji**: Okay, you told me that kitty has it documented why(?) they are evenly spaced. I did search for it and did not succed. Would you mind helping me out here? I would also love to forward you the discussion at #404 and its outcome at [TUC](https://github.com/contour-terminal/terminal-unicode-core). I would love to get your feedback on that, short or long. Then I would also love to very son PoC implement it (won't be hard at all) and would like to know if you have interest in playing around with that funcitonality to give us feedback on the usefulness for you. Ideally maybe also as a feature to be indicated in `notcurses-info` on top next to the others and depending on that, also (if enabled) the emoji line would then look evenly spaced, too!. Sounds interesting?
* quite a few more codepoints you use I do not manually construct yet. They'll be coming soon, too.

Okay, so much about just asking **a** simple question. Now this post looks like a status report spiced with a few questions. I hope I did not scare you or others away  with that. ;)

**Comment by dankamongmen** [comment]
> * you don't seem to believe that Contour can do undercurl, but we support various styled underlines as well as colorizing them (de-facto std CSI sequences)

export the `smulx` capability in your terminfo file, and i'll pick this up! i can add it as a heuristic but ideally you'd have `smulx` exported.

> * `3x2-` is indicated as not available. hm, I agree, I just merged these to master.

agreed, looking much better with `Contour 0.2.1-unreleased-master-8818676`. are you going to cut a 0.2.2 soon? if so, i'll put in a heuristic to give you `3x2` on 0.2.2+.

> * `defbg 0x101010 not considered transparent` How do you decide that something is or is not transparent? Am I maybe not passing all the information to you so that you can find out?

you don't want it considered transparent =]. kitty has a feature where the default background color is treated as transparent. i can understand kovid's reasoning for this, but it's a pain in the ass for me, where i want the background i ask for. the "not considered transparent" is a kitty-only thing, where whenever the default background color is used, i have to modify it a tiny bit to work around this behavior. a default background of 0x000000 means a chosen background of 0x000000 becomes 0x000001, for instance. 0xffffff becomes 0xfffffe. you don't want this.

> * smooth mosaics: I will be working on them next

not sure what you mean by this? in Notcurses parlance, a "mosaic" is a graphic broken down into many graphics each having cell size, but i don't yet do this. maybe you mean the diagonal unicode 13 symbols for legacy computing?

> * mathematical symbols (I so far only implemented square brackets I think, yesterday, rest will follow)

cool, no real concern to me save if users are using them; i don't use them myself.

> * you detect that sixels are supported but I do not see the tiny graphics in the top right anymore (was it ever graphics? I don't remember, but that fancy logo is missing)

presumably you're running a copy built from git? it's going to look for `notcurses.png` in `/usr/local/share/notcurses` unless you change the default installation prefix. this is done when building packages, and the packaged binary looks for `/usr/share/notcurses/notcurses.png`, which you probably have installed if you have installed a notcurses package. you can either run an installation from your git checkout, or `ln -s /usr/share/notcurses /usr/local/share/notcurses`.

you'll then see that Contour draws the image on a black background, annihilating the glyphs underneath, unlike other sixel implementations which interpret unspecified pixels as transparent =]. fixing *that* would be the thing that earned you the most high marks in my book =].

> * roundish box drawings and diagonal mosaics I will be adding soon.
> * **emoji**: Okay, you told me that kitty has it documented why(?) they are evenly spaced. I did search for it and did not succed. Would you mind helping me out here? I would also love to forward you the discussion at [VT Unicode Core Specification #404](https://github.com/contour-terminal/contour/issues/404) and its outcome at [TUC](https://github.com/contour-terminal/terminal-unicode-core). I would love to get your feedback on that, short or long. Then I would also love to very son PoC implement it (won't be hard at all) and would like to know if you have interest in playing around with that funcitonality to give us feedback on the usefulness for you. Ideally maybe also as a feature to be indicated in `notcurses-info` on top next to the others and depending on that, also (if enabled) the emoji line would then look evenly spaced, too!. Sounds interesting?

https://sw.kovidgoyal.net/kitty/faq/#some-special-symbols-are-rendered-small-truncated-in-kitty
https://github.com/kovidgoyal/kitty/issues/73
https://github.com/kovidgoyal/kitty/issues/1978

i will take a look at your doc. my current primary practical thought on this is https://github.com/dankamongmen/notcurses/issues/2199. https://github.com/dankamongmen/notcurses/issues/1329 has some thoughts. https://github.com/dankamongmen/notcurses/issues/1904 is not likely to be a solution to IMHO.

**Reply by christianparpart** [reply]
p.s.: sorry for the late response. I wanted to carefully reply to all points, not just loosely.

> > * you don't seem to believe that Contour can do undercurl, but we support various styled underlines as well as colorizing them (de-facto std CSI sequences)
> 
> export the `smulx` capability in your terminfo file, and i'll pick this up! i can add it as a heuristic but ideally you'd have `smulx` exported.

Hmm, what am I doing wrong?

<img width="1220" alt="image" src="https://user-images.githubusercontent.com/56763/136621130-95343ca6-5656-4124-8d2e-ee3cc327c524.png">

Isn't that how it should be? (don't do heuristics, i'd rather fix my own stupidity :) )

> > * `3x2-` is indicated as not available. hm, I agree, I just merged these to master.
> 
> agreed, looking much better with `Contour 0.2.1-unreleased-master-8818676`. are you going to cut a 0.2.2 soon? if so, i'll put in a heuristic to give you `3x2` on 0.2.2+.

💀 the heuristic :)
current release is `0.2.1-unreleased-....` that version number is below version `0.2.1` (semver.org). I will be releasing 0.2.1 as soon as possible. It's only bug fixes going in. Okay okay, and some box drawing additions, apparently, to make notcurses-info acting like a smiling face - but I do consider that a bugfix, too, in fact. :)


> > * `defbg 0x101010 not considered transparent` How do you decide that something is or is not transparent? Am I maybe not passing all the information to you so that you can find out?
> 
> you don't want it considered transparent =]. kitty has a feature where the default background color is treated as transparent. i can understand kovid's reasoning for this, but it's a pain in the ass for me, where i want the background i ask for. the "not considered transparent" is a kitty-only thing, where whenever the default background color is used, i have to modify it a tiny bit to work around this behavior. a default background of 0x000000 means a chosen background of 0x000000 becomes 0x000001, for instance. 0xffffff becomes 0xfffffe. you don't want this.

holy 💩 

> > * smooth mosaics: I will be working on them next
> 
> not sure what you mean by this? in Notcurses parlance, a "mosaic" is a graphic broken down into many graphics each having cell size, but i don't yet do this. maybe you mean the diagonal unicode 13 symbols for legacy computing?

U+1FB3C .. U+1FB6F 

I was about to implement as many as possible from https://www.unicode.org/charts/PDF/U1FB00.pdf and now I think I've come pretty far. 🏁 


> [...]
> > * you detect that sixels are supported but I do not see the tiny graphics in the top right anymore (was it ever graphics? I don't remember, but that fancy logo is missing)
> 
> presumably you're running a copy built from git? it's going to look for `notcurses.png` in `/usr/local/share/notcurses` unless you change the default installation prefix. this is done when building packages, and the packaged binary looks for `/usr/share/notcurses/notcurses.png`, which you probably have installed if you have installed a notcurses package. you can either run an installation from your git checkout, or `ln -s /usr/share/notcurses /usr/local/share/notcurses`.

you got me. sorry. yes. I'm running from latest git master from my home dir and nothing installed in that regard.

> you'll then see that Contour draws the image on a black background, annihilating the glyphs underneath, unlike other sixel implementations which interpret unspecified pixels as transparent =]. fixing _that_ would be the thing that earned you the most high marks in my book =].

Challenge excepted. 💣 I see what I can do to get it into the next bugfix cycle (release 0.2.1), as that can be seen as a fix too, because that's how xterm is rendering it, right?

> [...]
> > * **emoji**: Okay, you told me that kitty has it documented why(?) they are evenly spaced. I did search for it and did not succed. Would you mind helping me out here? I would also love to forward you the discussion at [VT Unicode Core Specification #404](https://github.com/contour-terminal/contour/issues/404) and its outcome at [TUC](https://github.com/contour-terminal/terminal-unicode-core). I would love to get your feedback on that, short or long. Then I would also love to very son PoC implement it (won't be hard at all) and would like to know if you have interest in playing around with that funcitonality to give us feedback on the usefulness for you. Ideally maybe also as a feature to be indicated in `notcurses-info` on top next to the others and depending on that, also (if enabled) the emoji line would then look evenly spaced, too!. Sounds interesting?
> 
> https://sw.kovidgoyal.net/kitty/faq/#some-special-symbols-are-rendered-small-truncated-in-kitty [kovidgoyal/kitty#73](https://github.com/kovidgoyal/kitty/issues/73) [kovidgoyal/kitty#1978](https://github.com/kovidgoyal/kitty/issues/1978)
> 
> i will take a look at your doc. my current primary practical thought on this is [dankamongmen/notcurses#2199](https://github.com/dankamongmen/notcurses/issues/2199). [dankamongmen/notcurses#1329](https://github.com/dankamongmen/notcurses/issues/1329) has some thoughts. [dankamongmen/notcurses#1904](https://github.com/dankamongmen/notcurses/issues/1904) is not likely to be a solution to IMHO.

did you have the time to take a look at that doc?

**Reply by dankamongmen** [reply]
> > export the `smulx` capability in your terminfo file, and i'll pick this up! i can add it as a heuristic but ideally you'd have `smulx` exported.
> Hmm, what am I doing wrong?

argh, my bad, `Smulx`. `smul` is for regular underline. `smulx` is not a thing, as far as i know. what i usually see is:

`Smulx=\E[4:%p1%dm`

which it looks like you have. https://github.com/dankamongmen/notcurses/blob/master/src/lib/termdesc.c#L655-L686

confirmed locally. looking at my Smulx code, it's kinda janky, and i think i've got a bug or two. the naive solution yielded up this:

![2021-10-08-183042_993x660_scrot](https://user-images.githubusercontent.com/143473/136631317-2e96dce8-515a-414f-9403-de41bc83e272.png)

i'll look into it this weekend

> <img alt="image" width="1220" src="https://user-images.githubusercontent.com/56763/136621130-95343ca6-5656-4124-8d2e-ee3cc327c524.png">

wow, that's really pretty! 

are you primarily running on macOS rather than Linux? ok, maybe that explains something i was hesitating to ask about. contour is...quite slow, at least on my Linux workstation and laptop. is it significantly faster on macOS?

> Isn't that how it should be? (don't do heuristics, i'd rather fix my own stupidity :) )

yep, i've checked your terminfo; this is my bug

> skull the heuristic :) current release is `0.2.1-unreleased-....` that version number is below version `0.2.1` (semver.org). I will be releasing 0.2.1 as soon as possible. It's only bug fixes going in. Okay okay, and some box drawing additions, apparently, to make notcurses-info acting like a smiling face - but I do consider that a bugfix, too, in fact. :)

i have to have a heuristic here, unfortunately, since there's no way to discover support for 3x2 drawing =\. if you want to introduce a means, i'll happily start supporting it.

> holy hankey

heh

more answers later

**Reply by christianparpart** [reply]
I use osx only if i have to. I am primarily using Linux. So i am good damn sorry if it is slow. And therefore i need to fix it .

Is this something you feel it just recently got slow or always.

Mind there is a branch where i am fully rewritng the internal grid purely for performance. But attention: do not try that yet. I will ping you later for that. :)

**Reply by christianparpart** [reply]
For the heuristic replacement, let me first try to talk to other TE devs so that it is at least 3 TEs trying to agree on something.

**Reply by j4james** [reply]
> > you'll then see that Contour draws the image on a black background, annihilating the glyphs underneath, unlike other sixel implementations which interpret unspecified pixels as transparent =]. fixing _that_ would be the thing that earned you the most high marks in my book =].
> 
> Challenge excepted. 💣 I see what I can do to get it into the next bugfix cycle (release 0.2.1), as that can be seen as a fix too, because that's how xterm is rendering it, right?

Note that transparency is dependent on the _Background Select_ parameter (P2) of the sixel escape sequence. The background should only be transparent if P2 is set to 1. Xterm doesn't handle this correctly, so please don't copy their behaviour exactly.

**Reply by dankamongmen** [reply]
as is usually the case, my learned colleague @j4james has the right of it here

**Reply by dankamongmen** [reply]
also, if you might be able to do a non-transparent Sixel more efficiently than P2=1 Sixel, know that i at least **do** use P2=0 when i have an entirely opaque sixel, so it's probably worth doing the different paths, ymmv. i know foot does this.

**Comment by dankamongmen** [comment]
oh wow and yes, contour looks MUCH better in `notcurses-info` now, wonderful!

**Reply by christianparpart** [reply]
Thanks man. I am not done yet. I will ping you then. :)

**Comment by dankamongmen** [comment]
how do i turn off your `Smulx`? you've got `rmul=\E[24m`, which doesn't seem to work, and `\e[4:0m` (used by some other terminals supporting `Smulx`) doesn't seem to work either:

[schwarzgerat](0) $ echo -e '\e[4:1merp\e[24mkljlksja'
*erp*kljlksja (italics represent underlined material)
[schwarzgerat](0) $ echo -e '\e[4:3merp\e[24mkljlksja'
_erpkljlksja_ (italics represent curly-underlined material)
[schwarzgerat](0) $ echo -e '\e[4:3merp\e[4:0mkljlksja'
_erpkljlksja_ (same)
[schwarzgerat](0) $

**Reply by christianparpart** [reply]
I'll fix it! Gimme a sec (or two).

**Reply by christianparpart** [reply]
Fixed on master that it should.

**Comment by dankamongmen** [comment]
no, it's always been slow for me. ought i be using some particular compile options? it's probable that i'm running a debug build if that's the default configuration.

**Reply by christianparpart** [reply]
Oh yes. `cmake -DCMAKE_BUILR_TYPE=RelWithDebInfo ..`

But there will also be major(!) Performance improvements imminent.

**Reply by dankamongmen** [reply]
ahh inded. i have the following in my `CMakeLists.txt`:

```c
if(NOT CMAKE_BUILD_TYPE AND NOT CMAKE_CONFIGURATION_TYPES)                                                                          
  set(CMAKE_BUILD_TYPE RelWithDebInfo CACHE                                                                                         
    STRING "Choose the build mode." FORCE)                                                                                          
endif()                                                                                                                             
set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS Debug Release MinSizeRel RelWithDebInfo Coverage)  
```

to default to `RelWithDebInfo`.

ok, yes, this is wholly acceptable now.

**Reply by dankamongmen** [reply]
oh yeah, much much much better. you might want to consider that snippet i posted above. i had been refraining from recommending contour due to the performance. this is much more competitive.

**Comment by christianparpart** [comment]
@dankamongmen btw, did you **ever** try notcurses (notcurses-info specifically) on osx? None of the TEs I tested do look as expected, not even close. Terminal.app being the shittiest amongst the gansters, but I start to question if that's an issue on the TE side or OSX/PTY-kernel side. not sure yet. wanna see screenshots?

**Reply by christianparpart** [reply]
This is iTerm:

<img width="996" alt="image" src="https://user-images.githubusercontent.com/56763/136672583-bfc6f502-927a-4674-a85b-6a0f45316753.png">

and kitty

<img width="1042" alt="image" src="https://user-images.githubusercontent.com/56763/136672699-0c5e19b6-ba6f-4fb0-97ea-e6f1e93579ac.png">

both latest versions on latest osx.

**Reply by christianparpart** [reply]
@dankamongmen dude, i think i've found it! you're using `wctombs()` and I bet that's not capable of speaking Unicode 14? (notcurses.h line 1987). I think it would be best to provide a self-implemented one?

**Reply by dankamongmen** [reply]
`wctombs()` ought get support as the OS adds it (this is actually Unicode 13 material, not 14). are you seeing these glyphs work anywhere else on macOS? i have tested things there, but i am not a macOS expert by any means, and my macOS test box is currently unavailable on top of that. =[

**Reply by dankamongmen** [reply]
but yes, i am thinking (for a number of reasons) that a replacement for POSIX `wcs*` is necessary, and have been for a long time (windows doesn't support it, for one).

**Reply by christianparpart** [reply]
@uspasojevic96 knows the most here. I am seeing this consistently because notcurses renders whitespace when wcstombs fails to convert, regardless of the terminal being used.

Echo'ing out the Codepoints directly works as expected.

**Reply by dankamongmen** [reply]
> Echo'ing out the Codepoints directly works as expected.

!!!

**Reply by kovidgoyal** [reply]
Dont ever use wcstombs. It is locale dependent. Terminals now-a-days all expect UTF-8. Encoding to UTF-8 is trivial, like a ten line function, feel free to use the one in charsets.c from the kitty source code.

**Reply by dankamongmen** [reply]
i'm always in a UTF-8 or US-ASCII encoding (enforced in `check_encoding()`, and `wcstombs()` is only dependent on `LC_CTYPE` (i.e. encoding). so i'm safe on that front. either way, i'm looking to get rid of it.

**Reply by kovidgoyal** [reply]
Good you are getting rid of it, but to be more precise wcstombs() is global state dependent. Since not curses is a library embedded in applications, there is no way you can enforce any value for that global state. A thread running in the application could change the locale just before the thread you are calling wcstombs in runs.

**Comment by christianparpart** [comment]
Hey @dankamongmen . I just git pull updated my local notcurses copy to run a fresh test and noticed you are having a TERMINALS.md. There however you mention that contour does NOT support Pixel `TIOCGWINSZ`?
I am pretty sure I was implementing this. Is there a accidental mistake on your side, or on my dark side?

**Reply by christianparpart** [reply]
hm, NOP, just checked. getting xpixel/ypixel via ioctl works just fine. :)

**Reply by christianparpart** [reply]
@dankamongmen ah, and `TERM=contour` is the correct one. The other one works, too and aught to always reflect the *latest* feature set, but I think I wasn't quite sane when I did this. So I'd suggest to just use `TERM=contour` and advise that.

WRT "Claims Sixel support."  Why do I only claim it? Don't I? What's missing? How can I change thy mind?

**Reply by dankamongmen** [reply]
> Hey @dankamongmen . I just git pull updated my local notcurses copy to run a fresh test and noticed you are having a TERMINALS.md. There however you mention that contour does NOT support Pixel `TIOCGWINSZ`? I am pretty sure I was implementing this. Is there a accidental mistake on your side, or on my dark side?

definitely an error on my part if so

**Reply by dankamongmen** [reply]
yeah it looks like the contour entry was written once and never updated, like an unloved terminfo database

**Reply by christianparpart** [reply]
like an unloved product.

**Reply by dankamongmen** [reply]
like a markdown that doesn't interact in any way with tests or day-to-day use. updated, thanks for the heads up!

and to be honest, until i started building with `ReleaseWithDebugInfo`, contour was very much in my second tier of terminals. `-O0` is no way to live.

**Reply by christianparpart** [reply]
I absolutely agree, `-O0` is unbearable. There I gladly stole your cmake part to default to `RelWithDebInfo`. 

With regards to performance, there is a lot of improvement to come. I'm trying to finally push my grid branch through that is fundamentally rewriting the underlying grid, currently trying to slim it down and leave all the optimizations that i did NOT finish stabelizing yet for future PRs. I'll buzz you later then.

With regards to usability (despite the tons of wishes I've already created tickets for), please just tell me. In the end I want to create a product that is "usable" - not just successfully compiling and used by me. :)

**Reply by dankamongmen** [reply]
i have found it completely usable now that it's not agonizingly slow, aside from when it crashes rendering glyphs uttered in the dreams of ten thousand dead drunkards. i promoted it into my first rank of terminals to test with, at the cost of one other which shall go unnamed, but really does need give me graphic transparency.

**Reply by dankamongmen** [reply]
and thanks i have updated several things in TERMINAL.md after a lookover

**Comment by xiongrenyuan** [comment]
您好，您的邮件已收到，祝好！（熊任远）
